src/test/java/spoon/test/issue4357/TestIssue4357.java

```diff
@@ -0,0 +1,33 @@
+package spoon.test.issue4357;
+
+import org.junit.jupiter.api.Test;
+import spoon.Launcher;
+import spoon.reflect.CtModel;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.nio.file.Paths;
+
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class TestIssue4357 {
+    
+    @Test
+    public void testClasspathURLWithSpaces() throws MalformedURLException {
+        // contract: launcher can handle spaces in classpath URL
+        Launcher launcher = new Launcher();
+        URL[] classpath = {
+                Paths.get("./src/test/resources/path with spaces/lib/bar.jar").toAbsolutePath().toUri().toURL()
+        };
+        launcher.getEnvironment().setNoClasspath(false);
+        launcher.getEnvironment().setShouldCompile(true);
+        ClassLoader classLoader = new URLClassLoader(classpath);
+        launcher.getEnvironment().setInputClassLoader(classLoader);
+        launcher.addInputResource(Paths.get("./src/test/resources/path with spaces/Foo.java").toAbsolutePath().toString());
+        CtModel model = launcher.buildModel();
+
+        assertTrue(model.getAllTypes().stream().anyMatch(ct -> ct.getQualifiedName().equals("Foo")), 
+                "CtTxpe 'Foo' not present in model");
```

Comment:
Optional change: This only indirectly verifies that we correctly handle spaces in classpath URLs. I'd suggest actually verifying that the reference to `bar.Bar` is correctly resolved.

```suggestion
        launcher.buildModel();

        CtType<?> foo = launcher.getFactory().Type().get("Foo");
        CtMethod<?> methodWithBarReturnType = foo.getMethod("foo");
        CtTypeReference<?> barRef = methodWithBarReturnType.getType();

        assertThat(barRef.getQualifiedName(), equalTo("bar.Bar"));
        assertThat(barRef.getTypeDeclaration(), notNullValue());
        assertTrue(barRef.getTypeDeclaration().isShadow(), "expected bar.Bar to be a shadow class");
```

Another alternative would be to test this with a unit test of `StandardEnvironment` instead, which is where the problem actually lies. Since this is tested so "far away" from the problem, the assertion should be very specific to IMO.